File: apps/haskell/design-pattern/src/TemplateMethod.hs
-- | Template Method パターン
-- レコードに関数フィールドを持たせ、アルゴリズムの骨格を固定しつつ
-- 各ステップを差し替え可能にする。
module TemplateMethod
  ( ReportFormat(..)
  , generateReport
  , htmlFormat
  , plainTextFormat
  ) where

-- | レポートの各ステップを関数で表現したレコード
data ReportFormat = ReportFormat
  { outputStart :: String -> String
  , outputLine  :: String -> String
  , outputEnd   :: String -> String
  }

-- | テンプレートメソッド: 骨格を固定し、各ステップを ReportFormat に委譲する
generateReport :: ReportFormat -> String -> [String] -> String
generateReport fmt title body =
  outputStart fmt title
  ++ concatMap (outputLine fmt) body
  ++ outputEnd fmt title

-- | HTML 形式
htmlFormat :: ReportFormat
htmlFormat = ReportFormat
  { outputStart = \t -> "<html><head><title>" ++ t ++ "</title></head><body>\n"
  , outputLine  = \l -> "  <p>" ++ l ++ "</p>\n"
  , outputEnd   = \_ -> "</body></html>\n"
  }

-- | プレーンテキスト形式
plainTextFormat :: ReportFormat
plainTextFormat = ReportFormat
  { outputStart = \t -> "***** " ++ t ++ " *****\n"
  , outputLine  = \l -> l ++ "\n"
  , outputEnd   = \_ -> ""
  }

File: apps/fsharp/design-pattern/src/DesignPattern/TemplateMethod.fs
namespace DesignPattern.TemplateMethod

/// Template Method パターン
/// F# では高階関数とレコード型で表現する。
/// 抽象クラスの代わりに、関数フィールドを持つレコード型を使う。
module TemplateMethod =

    /// レポートの出力形式を定義するレコード型
    type ReportFormat =
        { OutputStart: string -> string
          OutputHead: string -> string
          OutputBody: string list -> string list
          OutputEnd: string -> string }

    /// レポートを生成するテンプレートメソッド
    let generateReport (format: ReportFormat) (title: string) (items: string list) : string list =
        let startLine = format.OutputStart title
        let headLine = format.OutputHead title
        let bodyLines = format.OutputBody items
        let endLine = format.OutputEnd title
        [ startLine; headLine ] @ bodyLines @ [ endLine ]

    /// HTML フォーマット
    let htmlFormat: ReportFormat =
        { OutputStart = fun _ -> "<html>"
          OutputHead = fun title -> sprintf "  <head><title>%s</title></head>" title
          OutputBody = fun items -> items |> List.map (fun item -> sprintf "  <body>%s</body>" item)
          OutputEnd = fun _ -> "</html>" }

    /// プレーンテキストフォーマット
    let plainTextFormat: ReportFormat =
        { OutputStart = fun _ -> "***** Start *****"
          OutputHead = fun title -> title
          OutputBody = fun items -> items |> List.map id
          OutputEnd = fun _ -> "***** End *****" }
